sql-plugin/src/main/330+/scala/com/nvidia/spark/rapids/shims/v2/GpuHashPartitioning.scala

```diff
@@ -27,8 +27,12 @@ case class GpuHashPartitioning(expressions: Seq[Expression], numPartitions: Int)
   override def satisfies0(required: Distribution): Boolean = {
     super.satisfies0(required) || {
       required match {
-        case ClusteredDistribution(requiredClustering, _) =>
-          expressions.forall(x => requiredClustering.exists(_.semanticEquals(x)))
+        case c @ ClusteredDistribution(requiredClustering, requireAllClusterKeys, _) =>
+          if (requireAllClusterKeys) {
+            c.areAllClusterKeysMatched(expressions)
```

Comment:
Same question here. Do we have tests that cover this case? Because the reason we have to override it is so that it will play nicely with GPU expressions.

Reply:
I got lost inside test cases. I added debug messages inside `columnarEval` but was not able to pinpoint integration tests that cover that code.
At least `com.nvidia.spark.rapids.JoinsSuite` did.

`areAllClusterKeysMatched` is executing something like ` l.semanticEquals(r)` 
`SQLConf.REQUIRE_ALL_CLUSTER_KEYS_FOR_DISTRIBUTION` is false bydefault. Should I file a followup issue to create the tests needs for Range/hash partitioning with `REQUIRE_ALL_CLUSTER_KEYS_FOR_DISTRIBUTION` set to true?
Or do you have quick pointers for me to add basic tests within this PR?

Reply:
It is a little more complicated than just seeing that the code is called. I should have explained this in the original question. We do some odd things when replacing expressions. A simple attribute reference (like make sure the keys match column a) is fine. The problem shows up if we are doing computation as a part of that. So things like make sure the keys match column a + column b.  Because we replace the + with a GPU version things don't always work out. I don't know what `areAllClusterKeysMatched` does. So that is probably the first place to look. After that, yes we need tests to verify that this code is being executed and that it is doing the right thing, especially in cases when there is some kind of computation happening.

Reply:
`areAllClusterKeysMatched` in [ClusteredDistribution](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala#L103) checks that `for-all` expressions they semantically equal to the cluster keys. The previous code used to check `if exists`.

- One possible test is to have a new unit test GpuHashPartitioning then asserts on the return of `GpuHashPartitioning.satisfies` for a different sequence of expressions. But I was not yet sure if this is good enough and which expressions should we care about for the test.
-  Other way is to compare the captured execution plan to the expected. This is more complicated to me because I am not familiar yet with execution plans and where they would end up running GPU/CPU.

Reply:
Would it make sense to do the testing as a followup issue (still for 22.04)?  Currently the 3.3 builds are broken, and merging this would help unblock it.  We could unblock 3.3 builds in parallel with working on tests.  The code is doing the same thing wrt. `semanticEquals` that was done before, so if there's an issue there it's probably not specific to Spark 3.3.

Reply:
You are correct and I think we are probably okay with this change. I am just being paranoid. `semanticEquals` involves canonicalizing the expressions and then checking for equality. It should be fine. I am going to just call it good. I just have been bit in the past when only part of something is moved to the GPU, but not all of it. Because the current code works I am going to assume that the new code will work too.